### PR TITLE
Avoid mem copy when load data

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: https://github.com/milvus-io/milvus/issues/22986
/kind improvement

Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 1]
```diff
@@ -150,14 +158,66 @@ ScalarIndexSort<T>::Load(const BinarySet& index_binary, const Config& config) {
...
+        for (auto& item : meta_data[META]) {
+            std::string prefix = item[NAME];
+            if (prefix == index_data_key) {
+                data_.resize(index_size);
+                int slice_num = item[SLICE_NUM];
+                auto total_len = static_cast<size_t>(item[TOTAL_LEN]);
+
+                int64_t offset = 0;
+                for (auto i = 0; i < slice_num; ++i) {
+                    std::string file_name = prefix + "_" + std::to_string(i);
```
Comment (longjiquan) [inline internal/core/src/index/ScalarIndexSort-inl.h:185]: Make this a util function.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/core/src/common/Slice.cpp [hunk 2]
```diff
@@ -100,6 +94,69 @@ Disassemble(BinarySet& binarySet) {
...
+        if (data->size > slice_size_in_byte) {
+            slice_key_list.push_back(key);
+        } else if (key != INDEX_FILE_SLICE_META) {
+            ret.Append(key, data);
+        }
+    }
+
+    for (auto& key : slice_key_list) {
+        Config slice_i;
+        ShallowSlice(key, binarySet.GetByName(key), slice_size_in_byte, ret, slice_i);
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/common/Slice.cpp:144]: I think we need full test on this behavior.

File: internal/core/src/common/Slice.cpp [hunk 3]
```diff
@@ -100,6 +94,69 @@ Disassemble(BinarySet& binarySet) {
...
+    if (!data_src) {
+        return;
+    }
+
+    int slice_num = 0;
+    for (int64_t i = 0; i < data_src->size; ++slice_num) {
+        int64_t ri = std::min(i + slice_len, data_src->size);
+        auto deleter = [&](uint8_t*) {};  // avoid repeated deconstruction
+        auto slice_i = std::shared_ptr<uint8_t[]>(data_src->data.get() + i, deleter);
+        data_dst.Append(prefix + "_" + std::to_string(slice_num), slice_i, ri - i);
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/common/Slice.cpp:109]: And make this function inline.

File: internal/core/src/common/Slice.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/common/Slice.cpp:144]: Can you describe the lifetime of the memory in the binary set in detail? For example, when they are created, and who own them, and finally when and who free them?

File: internal/core/src/common/Slice.cpp [hunk 4]
```diff
@@ -100,6 +94,69 @@ Disassemble(BinarySet& binarySet) {
...
+        index_data_sliced = true;
+        auto slice_meta = binarySet.GetByName(INDEX_FILE_SLICE_META);
+        Config last_meta_data =
+            Config::parse(std::string(reinterpret_cast<char*>(slice_meta->data.get()), slice_meta->size));
+        for (auto& item : last_meta_data[META]) {
+            meta_info[META].emplace_back(item);
+        }
+    }
+
+    const int64_t slice_size_in_byte = index_file_slice_size << 20;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/common/Slice.cpp:132]: any other config?

File: internal/core/src/common/Slice.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/common/Slice.cpp:144]: It's safe to use shallowDisassemble in binary_set's lifecycle

File: internal/core/src/common/Slice.cpp [hunk 5]
```diff
@@ -100,6 +94,69 @@ Disassemble(BinarySet& binarySet) {
...
+        }
+    }
+
+    for (auto& key : slice_key_list) {
+        Config slice_i;
+        ShallowSlice(key, binarySet.GetByName(key), slice_size_in_byte, ret, slice_i);
+        meta_info[META].emplace_back(slice_i);
+        index_data_sliced = true;
+    }
+    if (!index_data_sliced) {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/common/Slice.cpp:148]: has been fixed

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 6]
```diff
@@ -150,14 +158,66 @@ ScalarIndexSort<T>::Load(const BinarySet& index_binary, const Config& config) {
...
 }
 
+template <typename T>
+void
+ScalarIndexSort<T>::AssembleAndLoadIndexDatas(const std::map<std::string, storage::FieldDataPtr>& index_datas) {
+    size_t index_size;
+    std::string index_length_key = "index_length";
+    AssertInfo(index_datas.find(index_length_key) != index_datas.end(), "loss index length file");
+    auto index_length_data = index_datas.at(index_length_key);
+    memcpy(&index_size, index_length_data->Data(), (size_t)index_length_data->Size());
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/index/ScalarIndexSort-inl.h:168]: Those code wited by you
https://github.com/milvus-io/milvus/blob/a182594bf5410a6c7e5f7261cb90a3688e0aab7a/internal/core/src/index/ScalarIndexSort-inl.h#L65-L83

File: internal/core/src/index/StringIndexMarisa.cpp [hunk 7]
```diff
@@ -188,13 +206,90 @@ StringIndexMarisa::Load(const BinarySet& set, const Config& config) {
...
+
+            if (prefix == MARISA_STR_IDS) {
+                str_ids_.resize(total_len / sizeof(size_t));
+                for (auto i = 0; i < slice_num; ++i) {
+                    std::string file_name = prefix + "_" + std::to_string(i);
+                    AssertInfo(index_datas.find(file_name) != index_datas.end(), "loss str_ids slice data");
+                    auto data = index_datas.at(file_name);
+                    int64_t len = data->Size();
+                    memcpy(reinterpret_cast<uint8_t*>(str_ids_.data()) + offset, data->Data(), len);
+                    offset += len;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/index/StringIndexMarisa.cpp:258]: It seems needed, please check

**Comment by xige-16** [comment]
rerun ut

File: internal/core/src/storage/Util.cpp [hunk 8]
```diff
@@ -336,30 +351,89 @@ EncodeAndUploadIndexSlice(RemoteChunkManager* remote_chunk_manager,
...
+
+// #else
+//     /* AIX, BSD, Solaris, and Unknown OS ------------------------ */
+//     return (size_t)0L; /* Unsupported. */
+// #endif
+// }
+
 std::vector<FieldDataPtr>
-GetObjectData(RemoteChunkManager* remote_chunk_manager, std::vector<std::string> remote_files) {
+GetObjectData(RemoteChunkManager* remote_chunk_manager, const std::vector<std::string>& remote_files) {
```
Comment (xiaofan-luan) [inline internal/core/src/storage/Util.cpp:399]: we can keep this function simple because threadpoll will keep the overall memory usage?

Comment (xiaofan-luan) [reply internal/core/src/storage/Util.cpp:399]: And I saw we already did protection in Memory File impl right

File: internal/core/src/storage/PayloadReader.cpp [hunk 9]
```diff
@@ -18,35 +18,67 @@
...
+    arrow::MemoryPool* pool = arrow::default_memory_pool();
+
+    // Configure general Parquet reader settings
+    auto reader_properties = parquet::ReaderProperties(pool);
+    reader_properties.set_buffer_size(4096 * 4);
+    reader_properties.enable_buffered_stream();
+
+    // Configure Arrow-specific Parquet reader settings
+    auto arrow_reader_props = parquet::ArrowReaderProperties();
+    arrow_reader_props.set_batch_size(128 * 1024);  // default 64 * 1024
```
Comment (xiaofan-luan) [inline internal/core/src/storage/PayloadReader.cpp:45]: any reason? 
for vector datas 128 *1024 seems to huge?
what about 16 * 1024

**Comment by xiaofan-luan** [comment]
My suggestion is to split the PR into 2 parts.
The storage format change looks great to me. 
The avoid Index assemble copy part looks a little bit complicated too me. We may want to keep the copy for now because current solution looks much cleaner. And vector index anyway need to be copied again

File: internal/core/src/storage/Util.cpp (hunk 8 above)
Comment (xige-16) [reply internal/core/src/storage/Util.cpp:399]: fixed

File: internal/core/src/storage/PayloadReader.cpp (hunk 9 above)
Comment (xige-16) [reply internal/core/src/storage/PayloadReader.cpp:45]: batch data size  not  larger than file size 16m

**Comment by xige-16** [comment]
rerun ut

**Pull request merged**